org.eclipse.xtext.ui/src/org/eclipse/xtext/ui/wizard/template/TemplateProcessor.xtend

```diff
@@ -34,22 +34,29 @@ import org.eclipse.xtend.lib.macro.file.Path
 @Beta
 abstract class TemplateProcessor extends AbstractClassProcessor {
 
+	// the annotation processors run in parallel but write the same file ... we synchronize to avoid race conditions 
+	static final Object LOCK = TemplateProcessor
```

Comment:
We may have to sync on some class object from the parent classloader instead. Should we add 'runSync' to the superclass as an abstract method?

Reply:
This would mean, we have two different class loaders that both load "TemplateProcessor" ... is this really a relevant case? Apart from this ... a lock that is potentially to weak is better then no lock, isn't it?

Reply:
I used poor mans debugging facilities `static Object LOCK = println(new Object);` and get this output in the console:
```
java.lang.Object@5e3c9d1a
java.lang.Object@3edab5a0
java.lang.Object@10d532e2
java.lang.Object@2e38dc6b
java.lang.Object@1e0f87cb
java.lang.Object@401d9784
```

So yes, there are plenty of classloaders involved in AA processing. Further debugging reveals, that we do use a single class loader per project that is build. So as long as we use the lock only in doGenerateCode, it should be ok, but suggests more safety than we can guarantee.